File: Looper.py
from enum import Enum
import Globals
import time

class Looper:
    class State(Enum):
        idle = 0
        first_recording = 1
        normal_operation = 2
        stopped = 3

    state = State.idle
    start_time = 0

    def __init__(self, rec_play_button, start_stop_button, reset_button, track_controllers):
        self.rec_play_button = rec_play_button
        self.start_stop_button = start_stop_button
        self.reset_button = reset_button
        self.buttons = [self.rec_play_button, self.start_stop_button, self.reset_button]
        self.track_controllers = []
        self.master_track = None
        self.track_length = 0
        self.track_controllers = track_controllers
        # self.state = 

    def tick(self):

        for track_controller in self.track_controllers:
            track_controller.tick()

        for button in self.buttons:
            button.tick()

        rec_play_button_pressed = self.rec_play_button.fell()
        start_stop_button_pressed = self.start_stop_button.fell()
        reset_button_pressed = self.reset_button.fell()
        
        if reset_button_pressed:
            print("Reset button pressed!")
            self.reset_pressed()

        if start_stop_button_pressed:
            if self.state is self.State.stopped:
                self.start_button()
                self.state = self.State.normal_operation
                self.start_time = time.time()
            else:
                self.stop_button()
                self.state = self.State.stopped

        #state action
        if self.state is self.State.idle:
            self.start_time = time.time()
        elif self.state is self.State.first_recording:
            pass
        elif self.state is self.State.normal_operation:
            # if Globals.master_done:
            #     Globals.master_done = False
            # if Globals.audio_wrap_around:
            #     Globals.audio_wrap_around = False
            #     Globals.master_done = True
                # self.start_time = time.time()
            if rec_play_button_pressed:
                print("record/play button pressed!")
                Globals.recording_mode = not Globals.recording_mode
        elif self.state is self.State.stopped:
            pass
        else:
            print("hit default state in looper state action")

        #state update
        if self.state is self.State.idle:
            for track_controller in self.track_controllers:
                if track_controller.is_recording():
                    self.master_track = track_controller
                    self.state = self.State.first_recording
                    print("first recording")
        elif self.state is self.State.first_recording:
            if self.master_track.is_playing():
                self.track_length = time.time() - self.start_time
                self.start_time = time.time()
                self.state = self.State.normal_operation
                print("normal operation")
        elif self.state is self.State.normal_operation:
            pass
        elif self.state is self.State.stopped:
            pass
        else:
            print("hit default state in looper state update")
                


    
    def addTrack(self, track):
        pass

    def test(self):
        return 2

    def reset_pressed(self):
        for track_controller in self.track_controllers:
            track_controller.reset_button()
        
        self.state = self.State.idle
        Globals.recording_mode = True
        self.master_track = None
        Globals.waiting_to_start = 0

    def start_button(self):
        for track_controller in self.track_controllers:
            track_controller.start_button()

    def stop_button(self):
        for track_controller in self.track_controllers:
            track_controller.stop_button()

File: Looper.cpp
/**
 * @file Looper.cpp
 * 
 * @brief Looper class
 * 
 * This file implements the Looper class
 * 
 * @author Bryan Cisneros
 */

#include "Looper.h"
#include "Globals.h"
#include "AudioInterface.h"
#include <stdio.h>

Looper::Looper(Button* recPlay, Button* startStop, Button* resetButton, Led* red_led, Led* green_led)
{
	// Point the member variables to the buttons and LEDs
	recPlayButton = recPlay;
	startStopButton = startStop;
	this->resetButton = resetButton;
	this->red_led = red_led;
	this->green_led = green_led;

	// Initialize to idle state, recording mode
	state = idle;
	recordingMode = true;
  	masterTrack = NULL;
}

void Looper::tick()
{
	// masterDone is set to true when a rollover (wrap-around) occurs. Rollover
	// is set by the audio interface, so we want to clear it as soon as possible,
	// but we want to make sure that masterDone is true for one (and only one)
	// whole tick process.
	masterDone = rollover;
	if (rollover)
	{
		rollover = false;
	}

	// Run the state machine for each track controller
	for (unsigned i = 0; i < trackControllers.size(); i++) {
		trackControllers[i]->tick();
	}

	// Run the state machine for each button
	recPlayButton->tick();
	startStopButton->tick();
	resetButton->tick();
  
	// Check if any of the buttons have been pressed
	bool recPlayButtonPressed = recPlayButton->fell();
	bool startStopButtonPressed = startStopButton->fell();
	bool resetButtonPressed = resetButton->fell();

	if (resetButtonPressed){
		resetPressed();
	}


	if (startStopButtonPressed){
		printf("Start/stop button pressed!\n");

		// If we were in the stopped state, call the startButton function.
		// If we were in any other state, call the stopButton function
		switch (state){
		case Looper::stopped:
			//start playing again
			startButton();

			//move to normal operation
			state = normalOperation;
			audio_set_track_position(0);

			break;
		default:
			stopButton();
			state = stopped;
			break;
		}
	}
  

	//state action
	switch (state){
  	case Looper::idle:
  		break;
  	case Looper::firstRecording:
		break;
  	case Looper::normalOperation:
		if (recPlayButtonPressed) {
			printf("recPlay button pressed!\n");
			recordingMode = !recordingMode;
			if (recordingMode)
			{
				// In recording mode, red should be on and green off
				green_led->turnOff();
				red_led->turnOn();
			}
			else
			{
				// In playing mode, red should be off and green on
				red_led->turnOff();
				green_led->turnOn();
			}
		}
  		break;
    case Looper::stopped:
      	break;
  	default:
  		break;
	}

	//state update
	switch (state)
	{
	case Looper::idle:
		// Check each track to see if any of them have moved into the recording
		// state. If they have, we have started our first recording. That track
		// becomes our master track, and we move into the first recording state
		for (unsigned i = 0; i < trackControllers.size(); i++) {
			if (trackControllers[i]->getState() == TrackController::recording) {
				masterTrack = trackControllers[i];
				state = firstRecording;
				printf("looper state: firstRecording\n");
				break;
			}
		}
		break;
	case Looper::firstRecording:
		if (masterTrack->getState() == TrackController::playing) {
			// If the master track has moved to the playing state, then our first
			// recording is complete. move to the normal operation state
			rollover = false; // make sure rollover is false before normal operation
			state = normalOperation;
			printf("looper state: normalOperation\n");
		}
		break;
	case Looper::normalOperation:
		break;
  	case Looper::stopped:
	  	break;
	default:
		break;
	}	
}

void Looper::addTrack(TrackController* track)
{
	// Just add the ttrack controller to the vector
	trackControllers.push_back(track);
}

void Looper::stopButton(){
  // Call the stop button function on each of the track controllers
  for (unsigned i = 0; i < trackControllers.size(); i++) {
    trackControllers[i]->stopButton();
  }
}

void Looper::startButton(){
	// Call the start button function on each of the track controllers
	for (unsigned i = 0; i < trackControllers.size(); i++) {
		trackControllers[i]->startButton();
	}
}

void Looper::resetPressed(){
	// Reset everything back to default!
	printf("Reset button pressed!\n");
	state = idle;
	recordingMode = true;
	masterTrack = NULL;
	waitingToStart = 0;

	// Call the reset button function on each of the track controllers
	for (unsigned i = 0; i < trackControllers.size(); i++) {
    	trackControllers[i]->resetButton();
  	}
}

